### PR TITLE
DM-12248 Chart options as a dialog

### PR DESCRIPTION
**Pull request by tgoldina**
https://jira.lsstcorp.org/browse/DM-12248

- Implemented chart options as a modal dialog as described in the ticket
- Added an option to remove active trace
- Added modal dialog

### CODE REVIEW

**Comment by tgoldina** [comment]
Thank you for finding legendgroup bug. ("legendgroup" is controlling which
traces should be hidden together.) It was based on activeTrace, I'll make
it unique.

The absolute position on ModalDialog is needed so that the dialog would be
visible when the scrollbar is present and the chart is down the page (see
http://localhost:8080/firefly/demo/ffapi-highlevel-charttest.html for
example)  I'd like the dialog to be be visible when chart options are
clicked.

2017-11-01 13:15 GMT-07:00 Loi Ly <notifications@github.com>:

**Comment by loitly** [comment]
I am off branch already, so I can't see it.  But, based on what I suggested, I think it would centered it on the screen instead of aligned to top.  It's not a big deal.  I am sure it's a personal preference.

**Comment by tgoldina** [comment]
Fixed the legendgroup bug and chart error handling. I have followed Loi's suggestion and made the dialog centered, when document.body is not scrollable. Please check it out.

**Comment by robyww** [comment]
Here are some tweeks to the CSS I think would make it look a little better

```
.ModalDialog__content {
    box-shadow: 2px 2px 3px rgba(0, 0, 0, 0.30);
    background-color: #E5E5E5;
    margin: 5px 3px;
    padding: 3px;
    border: 1px solid rgba(0,0,0, .75);
    border-radius: 3px;
}
```
I think the modal window should be a little darker.
```
.ModalWindow {
    position: absolute;
    top: 0;
    left: 0;
    background-color: rgba(0, 0, 0, 0.4);
    display: flex;
    align-items: center;
    justify-content: center;
}
```

**Comment by robyww** [comment]
BUG: When I attempt to use the pull down on the image tool bar, I get the following exception:
```
DropDownMenu.jsx:19 Uncaught TypeError: Cannot read property 'style' of null
    at placeDropDown (DropDownMenu.jsx:19)
    at DropDownMenuWrapper.componentDidMount (DropDownMenu.jsx:44)
    at ReactCompositeComponent.js:265
    at measureLifeCyclePerf (ReactCompositeComponent.js:75)
    at ReactCompositeComponent.js:264
    at CallbackQueue.notifyAll (CallbackQueue.js:76)
    at ReactReconcileTransaction.close (ReactReconcileTransaction.js:80)
    at ReactReconcileTransaction.closeAll (Transaction.js:206)
    at ReactReconcileTransaction.perform (Transaction.js:153)
    at ReactUpdatesFlushTransaction.perform (Transaction.js:140)
```

File: src/firefly/js/ui/ModalDialog.jsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+                        {this.props.children}
+                    </div>
+                </div>
+            </div>
+        );
+    }
+}
+
+function getDocHeight() {
+    let appEl = document.querySelector('div#App.rootStyle');
```
Comment (robyww) [inline src/firefly/js/ui/ModalDialog.jsx]: I think this line assumes the the div id is `app`.  The system allows for the root div. see- `FFEntryPoint.js line 54`

File: src/firefly/js/ui/ModalDialog.jsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    if (!appEl) {
+        appEl = document.body;
+    }
+    return Math.max(appEl.clientHeight, appEl.scrollHeight);
+}
+
+function getDocWidth() {
+    let appEl = document.querySelector('div#App.rootStyle');
```
Comment (robyww) [inline src/firefly/js/ui/ModalDialog.jsx]: Here too.

File: src/firefly/js/ui/PopupStoreConnection.jsx [hunk 3]
```diff
@@ -36,7 +37,7 @@ export class PopupStoreConnection extends PureComponent {
         var {visible}= this.state;
         if (!visible) return false;
         var {dialogId,popupPanel,requestOnTop,zIndex}= this.props;
-        return  React.cloneElement(popupPanel,
+        return  get(popupPanel, 'type.name') !== 'PopupPanel' ? popupPanel : React.cloneElement(popupPanel,
```
Comment (robyww) [inline src/firefly/js/ui/PopupStoreConnection.jsx]: I don't know why this line is necessary.  You could ignore the properties that are passed. (FYI- The requestToClose, dialogId, visible, property is to support the x, which we might need to add later).

File: src/firefly/js/ui/PopupUtil.jsx [hunk 4]
```diff
@@ -17,6 +18,28 @@ export const INFO_POPUP= 'InfoPopup';
...
 // ------------------------------------------------------------
 
+export const MODAL_DIALOG_ID = 'ModalDialog';
+
+/**
+ * Creates and shows the modal dialog.
+ * @param {string | object}  content can be a string or a react component
+ * @param {boolean} [show=true] show or hide this dialog
+ */
+export function showModal(content, show=true) {
```
Comment (robyww) [inline src/firefly/js/ui/PopupUtil.jsx]: This is great!

**Comment by tgoldina** [comment]
I have followed Trey's suggestion and used the fixed position for the modal as well as addressed the  other comments. Please check it out.

Indeed, fixed position looks much cleaner.

**Comment by robyww** [comment]
BUG: If i bring up charttest and click on the options dialog  before it is finished plotting i get an react exception and a normal exceptions

**Comment by loitly** [comment]
Working exactly in API as in application mode.  Good work!

**Pull request merged**